### PR TITLE
Typescript type

### PR DESCRIPTION
**Pull request by NejcZdovc**
**Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?**
- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the following... 
- Impact:
- Migration path for existing applications: 
- Github Issue(s) this is regarding:

**Other information**:
This resolves issue #3045.

### CODE REVIEW

**Comment by marvinhagemeister** [comment]
Great work! Out of curiosity, since I'm new to typescript: How do you run the tests? Is a simple compile enough?

**Comment by NejcZdovc** [comment]
@marvinhagemeister You can read more about [here](http://definitelytyped.org/guides/contributing.html#tests). You run test with a simple compile. It there are no errors like:
```
loader-test.ts(9,9): error TS2322: Type '{ entry: { alpha: string; beta: string; }; module: { loade1rs: { test: RegExp; loader: string; }[...' is not assignable to type 'Webpack'.
  Types of property 'module' are incompatible.
    Type '{ loade1rs: { test: RegExp; loader: string; }[]; }' is not assignable to type 'Module'.
      Object literal may only specify known properties, and 'loade1rs' does not exist in type 'Module'.
```

definition is ok 😄

**Comment by NejcZdovc** [comment]
@TheLarkInn can you please check out TODO's in the code and tell me what to insert there.

File: schemas/tests/agressive-merging-test.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+/// <reference path="webpack.d.ts" />
```
Comment (yaycmyk) [inline schemas/tests/agressive-merging-test.js]: `aggressive`

Comment (NejcZdovc) [inline schemas/tests/agressive-merging-test.js]: Fixed

**Comment by TheLarkInn** [comment]
@gajus can you also assist here.

File: schemas/webpack.d.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    type RuleSetUse = RuleSetUseItem | RuleSetUseItem[];
+    type RuleSetUseItem = string | RuleSetUseItemObject;
+    type RuleSetQuery = Object | string;
+}
+
+interface Webpack {
+    /*
+     Set the value of `require.amd` and `define.amd`.
+     */
+    amd?: any; // TODO missing type
```
Comment (zhaoshengjun) [inline schemas/webpack.d.ts]: - Might change the comment on properties to JSDoc style. This enable better tooltip in editor. For exmaple
```
/**
 *Set the value of `require.amd` and `define.amd`.
 */
````

Comment (zhaoshengjun) [inline schemas/webpack.d.ts]: `amd?: "require.amd" | "define.amd"`;

File: schemas/webpack.d.ts [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+     */
+    bail?: boolean;
+
+    /*
+     Cache generated modules and chunks to improve performance for multiple incremental builds.
+
+     Boolean: You can pass `false` to disable it.
+     Object: You can pass an object to enable it and let webpack use the passed object as cache.This way you can share the cache object between multiple compiler calls.
+     */
+    cache?: (boolean | Object);
```
Comment (zhaoshengjun) [inline schemas/webpack.d.ts]: Suggest to add an Object interface. Reason is the current definition will let `cache = "true"` pass, which should not.

```
interface Object {
[key: string]: any;
}
```

After adding this, the above statement will cause error.

File: schemas/webpack.d.ts [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+        unsafeCache?: boolean | Object;
+    }
+
+    interface ResolveAlias {
+        alias?: string;
+        name?: string;
+        onlyModule?: boolean;
+    }
+
+    type EntryItem = string | string[];
```
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: It's not used only in one placed, that's why I added it to type.

`entry: WebpackType.EntryItem | {[key: string]: WebpackType.EntryItem};`

File: schemas/webpack.d.ts (hunk 2 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: Added.

File: schemas/webpack.d.ts (hunk 3 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: You are correct. Fixed

File: schemas/webpack.d.ts (hunk 4 above)
Comment (zhaoshengjun) [inline schemas/webpack.d.ts]: My fault, pointed to the wrong line.

I mean for `OutputLibraryTarget` , `Stats` and `Target`.

**Comment by zhaoshengjun** [comment]
Awesome work!

**Comment by NejcZdovc** [comment]
I finished with this task, so I am just waiting for you guys to fill out the blanks (where is TODO) and check if everything is in order. I will remove WIP tag from title.

File: schemas/webpack.d.ts (hunk 4 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: Personally I prefer to have this long list of enums in separate variable, so it's a lot more readable for me.

File: schemas/webpack.d.ts [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+
+
+declare namespace WebpackType {
+    interface Node {
+        Buffer?: boolean;
+        __dirname?: boolean | "mock";
+        __filename?: boolean | "mock";
+        console?: boolean;
+        global?: boolean;
+        process?: boolean
```
Comment (Kovensky) [inline schemas/webpack.d.ts]: `Buffer`, `console` and `process` can also be `"mock"` or `"empty"`.

File: schemas/webpack.d.ts [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+        // The JSONP function used by webpack for async loading of chunks.
+        jsonpFunction?: string;
+
+        // If set, export the bundle as library. `output.library` is the name.
+        library?: string | string[];
+
+        libraryTarget?: OutputLibraryTarget;
+
+        // The output directory as **absolute path** (required).
+        path?: string;
```
Comment (Kovensky) [inline schemas/webpack.d.ts]: If it is required, it should not have a `?`.

File: schemas/webpack.d.ts [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+    }
+
+    interface OutputAuxiliaryComment {
+        amd?: string;
+        commonjs?: string;
+        commonjs2?: string;
+        root?: string;
+    }
+
+    interface OutputDevtoolLineToLine {
```
Comment (Kovensky) [inline schemas/webpack.d.ts]: Would be nice if this (and everything else using test/include/exclude) would work like a `RuleSetConditionObject`, but that's outside this scope

File: schemas/webpack.d.ts [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+        | "normal"
+        | "verbose";
+
+    type Target =
+        "web"
+        | "webworker"
+        | "node"
+        | "async-node"
+        | "node-webkit"
+        | "electron";
```
Comment (Kovensky) [inline schemas/webpack.d.ts]: `"atom"` / `"electron-main"` are (possibly?) aliases to `"electron"`. There is also a separate `"electron-renderer"` target.

File: schemas/webpack.d.ts [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+    /**
+     * TODO add description
+     */
+    devServer?: WebpackType.Object;
+
+    /**
+     * A developer tool to enhance debugging.
+     * Note: Boolean can be used only with value `false`.
+     */
+    devtool?: string | boolean;
```
Comment (Kovensky) [inline schemas/webpack.d.ts]: Would `string | false` work, or are constants in types only allowed for strings?

File: schemas/webpack.d.ts (hunk 5 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: In [scheme.json](https://github.com/webpack/webpack/blob/master/schemas/webpackOptionsSchema.json#L673) which is a base line for this PR it's says only boolean.

File: schemas/webpack.d.ts (hunk 6 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: That's correct, but in the tests that you can find in this repository, there is a lot of [example](https://github.com/webpack/webpack/blob/54aa3cd0d6167943713491fd5e1110b777336be6/test) without path in output object, so I am not so sure about it. But on [website](http://webpack.github.io/docs/configuration.html#output-path) it's required.

File: schemas/webpack.d.ts (hunk 7 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: I am not sure what you mean by that

File: schemas/webpack.d.ts (hunk 8 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: As described [here](http://webpack.github.io/docs/configuration.html#target) there are only these. Same is defined in `scheme.json`.

File: schemas/webpack.d.ts (hunk 9 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: Sadly no.

File: schemas/webpack.d.ts (hunk 6 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: For now let's say it's required.

**Comment by NejcZdovc** [comment]
@sokra There are some inconsistencies between `scheme.json` and [documentation](http://webpack.github.io/docs/configuration.html#node). What is the correct definition?

File: schemas/webpack.d.ts (hunk 2 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: @zhaoshengjun I am not so sure about these two strings anymore. If you checkout [documentation](http://webpack.github.io/docs/configuration.html#amd) you can see there is an object example.

File: schemas/webpack.d.ts (hunk 7 above)
Comment (Kovensky) [inline schemas/webpack.d.ts]: They both have `include`, `exclude` and `test` keys, but they don't necessarily seem to work the same. Would be nice if both the implementation and the object format were the same.

File: schemas/webpack.d.ts (hunk 8 above)
Comment (Kovensky) [inline schemas/webpack.d.ts]: I checked the source code itself; looks like an omission in the docs and `scheme.json`.

**Comment by TheLarkInn** [comment]
> @sokra In documentation there is <node buildin>: true, "mock", "empty" or false. Is this something that can be a single value for node property instead of object with options?

I would say we conform exactly with the JSON Validation Schema. If we need to change this, we'll need to change it in both places. I want to keep these things in sync as much as possible.

File: schemas/webpack.d.ts (hunk 9 above)
Comment (TheLarkInn) [inline schemas/webpack.d.ts]: @mhegazy @danielrosenwasser is there a better solution for this thank you can think of?

File: schemas/webpack.d.ts (hunk 8 above)
Comment (TheLarkInn) [inline schemas/webpack.d.ts]: Let's create an issue for this for the JSON Schema so that once we fix it there, we fix it in both places.

**Comment by ianks** [comment]
Can these typings be added to the package.json?

**Comment by TheLarkInn** [comment]
@ianks yes definitely. @NejcZdovc @zhaoshengjun add a new field called `"typings": ` and specify the path to that typings file. 

Here's an example PR of someone doing it. https://github.com/jeffbcross/angularfire2/commit/b3a5a3a66bb93e8d81591a4c0afbe88342415381

**Comment by NejcZdovc** [comment]
@TheLarkInn first I wanted to confirm location of definition file before adding it into package.json. So is the location correct or should we move it?

**Comment by TheLarkInn** [comment]
@danielrosenwasser is there any general recommendation in placement for type definitions files in a package, that would contradict where we are creating it now in this PR.

**Comment by DanielRosenwasser** [comment]
In my opinion, the "schema" type should just be exposed as part of the webpack library `.d.ts` files, so that users can write something like 

```ts
import webpack = require("webpack");

let config: webpack.WebpackConfiguration = {
   // ...
}

export = config;
```

**Comment by NejcZdovc** [comment]
@DanielRosenwasser  I am curious why not add it to DT, import it as `@type/` and add it to the `package.json` as typings. Then use it as `let config: Webpack = {`. 

AFAIK this is the only `.d.ts` file that is part of webpack library. @TheLarkInn will there be more `.d.ts` files for example for plugins?

**Comment by TheLarkInn** [comment]
I think overall in the future if we transition this project to Typescript then each plugin technically will have a type. But for now we do not. I don't know if that answers the question for you.

**Comment by TheLarkInn** [comment]
#3112 had been merged so you can update accordingly

File: schemas/webpack.d.ts (hunk 5 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: It's fixed now.

File: schemas/webpack.d.ts (hunk 8 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: Done and added to the definition.

File: schemas/webpack.d.ts (hunk 2 above)
Comment (TheLarkInn) [inline schemas/webpack.d.ts]: Starting tracing the use of the config value from here: https://github.com/webpack/webpack/blob/master/lib/WebpackOptionsApply.js#L248

To: 
https://github.com/webpack/webpack/blob/master/lib/dependencies/AMDPlugin.js#L92-L100

It appears that the default is an empty Object, and that it is `{[key: string]: string}`

**Comment by NejcZdovc** [comment]
Updated accordingly of the #3112

File: schemas/webpack.d.ts (hunk 2 above)
Comment (NejcZdovc) [inline schemas/webpack.d.ts]: Fixed

File: schemas/webpack.d.ts [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+
+    resolveLoader?: WebpackType.Resolve;
+
+    /**
+     * Used by the webpack CLI program to pass stats options.
+     */
+    stats?: WebpackType.Object | boolean | WebpackType.Stats;
+
+    /**
+     * TODO add description
```
Comment (TheLarkInn) [inline schemas/webpack.d.ts]: "Specifies webpack deployment target. This modifies how the webpack bootstrap function is generated based on each target. Can also expose extra global variables into the webpack module scope (required for specific targets (IE: `electron`))

Comment (NejcZdovc) [inline schemas/webpack.d.ts]: > Specifies webpack deployment target. This modifies how the webpack bootstrap function is generated based on each target. Can also use predefined bootstrap functions based on targets such as `electron`, `atom`, etc.

What about something like that?

Comment (NejcZdovc) [inline schemas/webpack.d.ts]: As far as I understand this is that you can specify string (predefined function as `electron`) or you can specify completely custom function, which is called [here](https://github.com/webpack/webpack/blob/master/lib/WebpackOptionsApply.js#L184).

That' why we have: `target?: Function | WebpackType.Target;`

File: schemas/webpack.d.ts (hunk 9 above)
Comment (DanielRosenwasser) [inline schemas/webpack.d.ts]: `string | false` should be possible in TS 2.0

**Comment by NejcZdovc** [comment]
Ok, so AFAIK this is blocking merge of this PR

- [ ] Position of this `d.ts` file
- [ ] How will we use TS in reality (for now we have two proposals)

How can I help to resolve this to, so I can close this and go to another issue? @TheLarkInn @DanielRosenwasser

File: schemas/webpack.d.ts [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+
+    type RuleSetUse = RuleSetUseItem | RuleSetUseItem[];
+    type RuleSetUseItem = string | RuleSetUseItemObject;
+
+    type RuleSetQuery = Object | string;
+
+    type RuleSetRules = RuleSetRule[];
+}
+
+interface Webpack {
```
Comment (niieani) [inline schemas/webpack.d.ts]: This interface should be exported, no? Doesn't make much sense having a type definition otherwise.

**Comment by SimenB** [comment]
Any plans to add matching flow type definitions?

**Comment by sokra** [comment]
old, controversal, there is typing in DT, closing

**Pull request closed without merging**